[RFC] Prevent stalls due to large tablet mutations

bhalevy opened the pull request:
Currently, replica::tablet_map_to_mutation generates a mutation having a row per tablet.
With enough tablets (10s of thousands) in the table we observe reactor stalls when freezing / unfreezing such large mutations, as seen in https://github.com/scylladb/scylladb/pull/18095#issuecomment-2029246954, and I assume we would see similar stalls also when converting those mutation into canonical_mutation and back, as they are similar to frozen_mutation, and bit more expensive since they also save the column mappings.

This series takes a different approach than allowing freeze to yield.
`tablet_map_to_mutation` is changed to `tablet_map_to_mutations`, able to generate multiple split mutations, that when squashed together are equivalent to the previously large mutation.  Those mutations are fed into a `process_mutation` callback function, provided by the caller, which may add those mutation to a vector for further processing, and/or process them inline by freezing or making a canonical mutation.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7870/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7870/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/839/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8869/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7870/)
#### Build Details:
- Duration: 2 hr 30 min
- Builder: spider4.cloudius-systems.com

bhalevy commented:
- Made process_func synchronous since even if and when we make freeze async it will require a thread so process_func returning a future won't help us freeze it gently.  We'd need to run everything, including tablet_map_to_mutations in a thread.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7875/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7875/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8874/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/844/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7875/)
#### Build Details:
- Duration: 2 hr 17 min
- Builder: spider4.cloudius-systems.com

tgrabiec commented on service/topology_coordinator.cc:
```diff
@@ -966,12 +966,15 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
-                guard.write_timestamp()));
+                guard.write_timestamp(),
+                [&] (mutation m) {
+                    out.emplace_back(m);
```
nit: std::move(m).

bhalevy replied:
`out` is a `std::vector<canonical_mutation>` so we're indirectly invoking `canonical_mutation(const canonical_mutation&)`, so std::move isn't needed.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1189,7 +1189,9 @@ class tablet_allocator_impl : public tablet_allocator::impl
-            muts.emplace_back(tablet_map_to_mutation(map, s.id(), s.keypace_name(), s.cf_name(), ts).get());
+            tablet_map_to_mutations(map, s.id(), s.keypace_name(), s.cf_name(), ts, [&] (mutation m){
+                muts.emplace_back(std::move(m));
+            }).get();
```
I think read_tablet_mutations() is also important to treat because that's when we read the whole tablet metadata.

bhalevy replied:
make sense. will add that as well.

bhalevy commented:
In v2 ([10b0e6a](https://github.com/scylladb/scylladb/commit/10b0e6a2344594a00703267a7d18e8498243a094))
- rebased
- mutation: add for_each_split_mutation
- storage_service: merge_topology_snapshot: apply in background
- frozen_mutation: add unfreeze_and_split_gently
- storage_service: merge_topology_snapshot: freeze snp.mutations gently
- tablets: read_tablet_mutations: use unfreeze_and_split_gently
- schema_tables: convert_schema_to_mutations: get process_mutation function
- schema_tables: convert_schema_to_mutations: simplify check for system keyspace

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7984/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7984/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/951/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8979/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7984/)
#### Build Details:
- Duration: 2 hr 26 min
- Builder: spider7.cloudius-systems.com

denesb commented:
@bhalevy is this ready for review/merge? Consider dropping `[RFC]` from the subject if so.

bhalevy commented:
> @bhalevy is this ready for review/merge?

Yes, it is.

> Consider dropping `[RFC]` from the subject if so.

Done.

tgrabiec commented on db/schema_tables.cc:
```diff
@@ -830,29 +831,21 @@ future<table_schema_version> calculate_schema_digest(distributed<service::storag
-    auto map = [&proxy, features] (sstring table) -> future<std::vector<canonical_mutation>> {
-        auto& db = proxy.local().get_db();
+    auto& db = proxy.local().get_db();
+    for (const auto& table : all_table_names(features)) {
```
There is loss of parallelism here, map_reduce() maps in parallel.

bhalevy replied:
What parallelism? unfreezing_gently is cpu-bound, preemption would be due to maybe_yield, so I don't see what that buys us.

tgrabiec replied:
Parallelism of db::system_keyspace::query_mutations(), which may go to disk.

bhalevy replied:
ok

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -752,35 +752,38 @@ future<> storage_service::merge_topology_snapshot(raft_snapshot snp) {
+                co_await _db.local().apply(s, m, {}, db::commitlog::force_sync::yes, db::no_timeout);
+            };
+            auto apply_in_background = [&] (frozen_mutation m) {
+                done = done.then([&, m = std::move(m)] () mutable {
```
Yeah :(

I also don't understand, how is this better from calling `max_concurrent_for_each` in the foreground? The previous code was simpler. The commit message does not explain how doing this background thing (but then actually waiting for it) is an improvement.

kbr-scylla replied:
Ah, so we basically replace a vector of frozen_mutations, with a list of tasks (`then` continuation chain). What's wrong with the vector though? The vector should be small anyway, even if it's "long" (but it won't ever be that long, really) because mutations are by themselves small objects IIUC (in the sense that they are tree-based data structures, small pointers to larger objects).

The complexity is IMO not worth it, and error prone (e.g. accidentally making everything sequential)

bhalevy replied:
The question is how many cdc mutations can we have and how much external memory they hold, not the size of the vector capacity per-se.
I'm concerned about a large number of tables with cdc enabled.

bhalevy replied:
> Since we wait for previous "done", there is no parallelism, mutations are applied sequentially.
> This can increase latency of snapshot application. \cc @kbr-scylla

The intention was the call `apply_mutation` and enqueuing only the returned future.
That said, we can consider this as a follow-up.
It's not part of the core changes of this PR.

kbr-scylla commented:
I dequeued the PR. I think the comments from @tgrabiec point to a serious issue.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7999/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7999/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7999/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8997/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/970/)
#### Failed Tests (2943/29478):
- [test_auth_no_quorum](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum)
- [test_auth_raft_snapshot_transfer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer)
- [test_auth_no_quorum](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum)
- [test_auth_raft_snapshot_transfer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer)
- [test_auth_no_quorum](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum)
- [test_auth_raft_snapshot_transfer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer)
- [test_auth_raft_command_split](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split)
- [test_auth_raft_command_split](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split)
- [test_auth_raft_command_split](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split)
- [test_auth_v2_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration)
#### Build Details:
- Duration: 5 hr 46 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: ABORTED
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7999/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7999/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7999/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8997/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/970/)
> 
> #### Failed Tests (2943/29478):
> * [test_auth_no_quorum](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum)
> * [test_auth_raft_snapshot_transfer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer)
> * [test_auth_no_quorum](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum)
> * [test_auth_raft_snapshot_transfer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer)
> * [test_auth_no_quorum](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum)
> * [test_auth_raft_snapshot_transfer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer)
> * [test_auth_raft_command_split](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split)
> * [test_auth_raft_command_split](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split)
> * [test_auth_raft_command_split](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split)
> * [test_auth_v2_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7999/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration)
> 
> #### Build Details:
> * Duration: 5 hr 46 min
> * Builder: spider1.cloudius-systems.com

Wow, there's a quite total meltdown of topology tests in debug mode (with address in use errors and mismatching cluster IDs) and about half of the dtests in gating-dtest-release-with-consistent-topology-changes.

bhalevy commented:
In v4 ([178feeb](https://github.com/scylladb/scylladb/commit/178feeb24ad99d4373a91956d26618c134f4ff65)):
- rebased
- dropped controversial / out-of-scope patches:
  - "storage_service: merge_topology_snapshot: apply in background"
  - "schema_tables: convert_schema_to_mutations: get process_mutation function"
- schema_tables: convert_schema_to_mutations: simplify check for system keyspace
  - added vector reserve

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8006/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8006/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/976/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9002/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8006/)
#### Build Details:
- Duration: 2 hr 27 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
6f83e59: rebased

But I'm converting this PR back to draft since I'm not sure we need all of it
after https://github.com/scylladb/scylladb/pull/18290

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9029/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9029/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1661/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9685/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9029/)
#### Failed Tests (76/30556):
- [test_node_failure_during_tablet_migration[streaming-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_source_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[streaming-source])
- [test_node_failure_during_tablet_migration[streaming-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_destination_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[streaming-destination])
- [test_node_failure_during_tablet_migration[allow_write_both_read_old-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_allow_write_both_read_old_source_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[allow_write_both_read_old-source])
- [test_node_failure_during_tablet_migration[write_both_read_old-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_write_both_read_old_source_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[write_both_read_old-source])
- [test_node_failure_during_tablet_migration[write_both_read_new-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_write_both_read_new_source_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[write_both_read_new-source])
- [test_node_failure_during_tablet_migration[write_both_read_new-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_write_both_read_new_destination_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[write_both_read_new-destination])
- [test_node_failure_during_tablet_migration[use_new-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_use_new_destination_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[use_new-destination])
- [test_node_failure_during_tablet_migration[cleanup-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_cleanup_source_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[cleanup-source])
- [test_node_failure_during_tablet_migration[cleanup_target-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_cleanup_target_destination_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[cleanup_target-destination])
- [test_node_failure_during_tablet_migration[end_migration-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9029/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_end_migration_source_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[end_migration-source])
#### Build Details:
- Duration: 4 hr 22 min
- Builder: i-024f3b4c57852b6f7 (i4i.8xlarge)

swasik commented:
> [6f83e59](https://github.com/scylladb/scylladb/commit/6f83e590333f4e82c84c0b79037f76fd4339e5ea): rebased
> 
> But I'm converting this PR back to draft since I'm not sure we need all of it after #18290

@bhalevy The last status update is from few months ago - do you think that we still need it? Do we have a plan for it?

bhalevy commented:
> > [6f83e59](https://github.com/scylladb/scylladb/commit/6f83e590333f4e82c84c0b79037f76fd4339e5ea): rebased
> > But I'm converting this PR back to draft since I'm not sure we need all of it after #18290
> 
> @bhalevy The last status update is from few months ago - do you think that we still need it? Do we have a plan for it?

Yes, I think we still need it with a very large number of tablets.
I'll rebase the branch.

bhalevy commented:
rebased

mykaul commented:
@bhalevy - please resolve conflicts.
